### ISSUE TITLE
request for comprehensive documentation and examples for library usage

### ISSUE THREAD

**Issue by inboxsphere** [SeekStorm/SeekStorm#15]
when using the library, there is a lack of necessary documentation. additionally, some types, such as SchemaField, cannot be constructed directly. It would be helpful to provide some examples that cover various use cases, such as index creation, writing data, searching, statistical analysis, pagination, etc. Otherwise, users will need to analyze the code in order to understand how to call the relevant functions.

**Comment by wolfgarbe** [comment]
You are right, there is much room for improvement. But unfortunately, the day has only 24 hours. This is what we have got so far:

**Client**

- [Library docs](https://docs.rs/seekstorm/latest/seekstorm/)
- [Usage examples](https://github.com/SeekStorm/SeekStorm?tab=readme-ov-file#documentation)
- [Examples faceted search](https://github.com/SeekStorm/SeekStorm?tab=readme-ov-file#faceted-search---quick-start)

**Server**

- [REST API documentation](https://seekstorm.apidocumentation.com/reference)
- [Examples](https://github.com/SeekStorm/SeekStorm/blob/main/src/seekstorm_server/README.md)
- [Interactive examples](https://github.com/SeekStorm/SeekStorm/blob/main/src/seekstorm_server/test_api.rest) (use with VSC extension "Rest client" to execute API calls and inspect responses)
- [Build a Wikipedia search engine with the SeekStorm server](https://github.com/SeekStorm/SeekStorm?tab=readme-ov-file#build-a-wikipedia-search-engine-with-the-seekstorm-server)
- [Build a PDF search engine with the SeekStorm server](https://github.com/SeekStorm/SeekStorm?tab=readme-ov-file#build-a-pdf-search-engine-with-the-seekstorm-server)

We will improve it over time, in parallel to the ongoing development.

**Comment by wolfgarbe** [comment]
>  I didn't know how to use advanced queries, such as returning all matching queries. When I used Length=int.Max in the query, the server crashed, and 0 resulted in no results. 

Currently, there are limitations on the maximum number of returned results, which is not yet documented, and exceeding the limit is not prevented. 

Here is the technical background: Top-k results are collected and sorted by score in descending order. To achieve this, all matching candidates are inserted into a min-heap structure. The top-K most relevant candidates are kept in the min-heap. The length parameter (=Top-K) determines the size of the min-heap structure and its underlying array. If you set Length=int.Max, that means that the min-heap array is initialized for 2 billion elements. And that usually doesn’t fit in RAM.

There are several ways to fix this. We could introduce (and document) an explicit limit for the length parameter. This is challenging if we want to use the maximum available RAM, which is different on every system. 

Another possibility is that instead of initializing the min-heap to length, we could incrementally grow the min-heap structure while candidates are inserted. This would help in cases where the number of actual found results is << length. But incrementally increasing capacity comes with a latency penalty.

But for billion-scale indices, which SeekStorm supports, it still won’t be possible to return all documents within a single search, in a single result set, only page-wise. Due to the inherent RAM limitation, we could overcome this only if we forgo the requirement that results should be ordered by score. Then we could bypass the min-heap structure if we want to return ALL matching results, but unordered.

>  I also couldn't find examples for + and -, 

*Query operators and query type*

Boolean queries are specified in the search method either via the query_type parameter or via operator chars within the query parameter.  
The interpretation of operator chars within the query string (set `query_type=QueryType::Union`) allows to specify advanced search operations via a simple search box.

Intersection, AND `+`
```rust ,no_run
use seekstorm::search::QueryType;
let query_type=QueryType::Union; 
let query="+red +apple".to_string();
```

```rust ,no_run
use seekstorm::search::QueryType;
let query_type=QueryType::Intersection; 
let query="red apple".to_string();
```

Union, OR
```rust ,no_run
use seekstorm::search::QueryType;
let query_type=QueryType::Union; 
let query="red apple".to_string();
```

Phrase `""`
```rust ,no_run
use seekstorm::search::QueryType;
let query_type=QueryType::Union; 
let query="\"red apple\"".to_string();
```

```rust ,no_run
use seekstorm::search::QueryType;
let query_type=QueryType::Phrase; 
let query="red apple".to_string();
```

Except, minus, NOT `-`
```rust ,no_run
use seekstorm::search::QueryType;
let query_type=QueryType::Union; 
let query="apple -red".to_string();
```

Mixed phrase and intersection
```rust ,no_run
use seekstorm::search::QueryType;
let query_type=QueryType::Union; 
let query="+\"the who\" +uk".to_string();
```

Query operator examples have been added to README.md and the library documentation

> or examples starting with xxx.

SeekSorm currently does not support prefix search
